consensus/consensus_v2.go

```diff
@@ -290,16 +297,19 @@ func (consensus *Consensus) onAnnounce(msg *msg_pb.Message) {
 
 // tryPrepare will try to send prepare message
 func (consensus *Consensus) prepare() {
-	// Construct and send prepare message
-	msgToSend := consensus.constructPrepareMessage()
-	// TODO: this will not return immediatey, may block
-
-	if err := consensus.msgSender.SendWithoutRetry([]nodeconfig.GroupID{nodeconfig.NewGroupIDByShardID(nodeconfig.ShardID(consensus.ShardID))}, host.ConstructP2pMessage(byte(17), msgToSend)); err != nil {
-		utils.Logger().Warn().Err(err).Msg("[OnAnnounce] Cannot send prepare message")
-	} else {
-		utils.Logger().Info().
-			Str("blockHash", hex.EncodeToString(consensus.blockHash[:])).
-			Msg("[OnAnnounce] Sent Prepare Message!!")
+	groupID := []nodeconfig.GroupID{nodeconfig.NewGroupIDByShardID(nodeconfig.ShardID(consensus.ShardID))}
+	for i, key := range consensus.PubKey.PublicKey {
+		// Construct and send prepare message
+		msgToSend := consensus.constructPrepareMessage(key, consensus.priKey.PrivateKey[i])
+		// TODO: this will not return immediatey, may block
+
+		if err := consensus.msgSender.SendWithoutRetry(groupID, host.ConstructP2pMessage(byte(17), msgToSend)); err != nil {
```

Comment:
So we are not grouping the messages together in a single message?

Reply:
Should be possible, but requires more changes to signing and verifying. I was planning to consider this for the open staking multi key support.